lib/wallet/src/Cardano/Wallet/Write/Tx.hs

```diff
@@ -0,0 +1,481 @@
+
+-- | For convenience working with 'IsRecentEra'. Similar to 'Cardano.cardanoEra,
+-- but with a 'IsRecentEra era' constraint instead of 'Cardano.IsCardanoEra.
+cardanoEra :: forall era. IsRecentEra era => Cardano.CardanoEra era
+cardanoEra = cardanoEraFromRecentEra $ recentEra @era
+
+-- | For convenience working with 'IsRecentEra'. Similar to
+-- 'Cardano.shelleyBasedEra, but with a 'IsRecentEra era' constraint instead of
+-- 'Cardano.IsShelleyBasedEra'.
+shelleyBasedEra :: forall era. IsRecentEra era => Cardano.ShelleyBasedEra era
+shelleyBasedEra = shelleyBasedEraFromRecentEra $ recentEra @era
+
+data InAnyRecentEra thing where
+     InAnyRecentEra :: IsRecentEra era -- Provide class constraint
+                    => RecentEra era   -- and explicit value.
+                    -> thing era
+                    -> InAnyRecentEra thing
+
+withRecentEra
+    :: InAnyRecentEra thing
+    -> (forall era. IsRecentEra era => thing era -> a)
+    -> a
+withRecentEra (InAnyRecentEra _era tx) f
+    = f tx
+
+-- | "Downcast" something existentially wrapped in 'Cardano.InAnyCardanoEra'.
+asAnyRecentEra
+    :: Cardano.InAnyCardanoEra a
+    -> Maybe (InAnyRecentEra a)
+asAnyRecentEra = \case
+    Cardano.InAnyCardanoEra Cardano.BabbageEra a ->
+        Just $ InAnyRecentEra RecentEraBabbage a
+    Cardano.InAnyCardanoEra Cardano.AlonzoEra a ->
+        Just $ InAnyRecentEra RecentEraAlonzo a
+    _ -> Nothing
+
+--------------------------------------------------------------------------------
+-- TxIn
+--------------------------------------------------------------------------------
+
+type TxIn = Ledger.TxIn StandardCrypto
+
+-- | Useful for testing
+unsafeMkTxIn :: ByteString -> Word -> TxIn
+unsafeMkTxIn hash ix = Ledger.mkTxInPartial
+    (toTxId hash)
+    (fromIntegral ix)
+  where
+    toTxId :: ByteString -> Ledger.TxId StandardCrypto
+    toTxId h =
+        (Ledger.TxId (unsafeMakeSafeHash $ UnsafeHash $ toShort h))
+
+--------------------------------------------------------------------------------
+-- TxOut
+--------------------------------------------------------------------------------
+
+type TxOut era = Core.TxOut era
+
+type Address = Ledger.Addr StandardCrypto
+
+unsafeAddressFromBytes :: ByteString -> Address
+unsafeAddressFromBytes bytes = case Ledger.deserialiseAddr bytes of
+    Just addr -> addr
+    Nothing -> error "unsafeAddressFromBytes: failed to deserialise"
+
+scriptFromBytes
+    :: ByteString
+    -> Either CBOR.DecoderError (Script LatestLedgerEra)
+scriptFromBytes bs = CBOR.decodeAnnotator "Script" CBOR.fromCBOR (fromStrict bs)
+
+scriptFromCardanoScriptInAnyLang
+    :: Cardano.ScriptInAnyLang
+    -> (Script LatestLedgerEra)
+scriptFromCardanoScriptInAnyLang
+    = Cardano.toShelleyScript
+    . fromMaybe (error "all valid scripts should be valid in latest era")
+    . Cardano.toScriptInEra latestEra
+  where
+    latestEra = Cardano.BabbageEra
```

Comment:
Question: are there round-trip tests for:
- `script{From,To}CardanoScriptInAnyLang`?
- `script{From,To}CardanoEnvelopeJSON`?

Reply:
Now, yes:

https://github.com/input-output-hk/cardano-wallet/blob/d0a7ef77bba61dddb1928c918b911ef2521a5d78/lib/wallet/test/unit/Cardano/Wallet/Write/TxSpec.hs#L125-L139